catalog/app/containers/Bucket/PackageDialog/PackageDialog.js

```diff
@@ -93,25 +93,61 @@ const readFile = (file) =>
     reader.readAsText(file)
   })
 
+const validateName = (req) =>
+  cacheDebounce(async (name) => {
+    if (name) {
+      const res = await req({
+        endpoint: '/package_name_valid',
+        method: 'POST',
+        body: { name },
+      })
+      if (!res.valid) return 'invalid'
+    }
+    return undefined
+  }, 200)
+
 export function useNameValidator() {
   const req = APIConnector.use()
+  const [counter, setCounter] = React.useState(0)
+  const [processing, setProcessing] = React.useState(false)
+  const inc = React.useCallback(() => setCounter(R.inc), [setCounter])
+
+  const validator = React.useMemo(() => validateName(req), [req])
+
+  const validate = React.useCallback(
+    async (name) => {
+      setProcessing(true)
+      const error = await validator(name)
+      setProcessing(false)
+      return error
+    },
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    [counter, validator],
+  )
+
+  return React.useMemo(() => ({ validate, processing, inc }), [validate, processing, inc])
+}
+
+export function useNameExistence(bucket) {
   const [counter, setCounter] = React.useState(0)
   const inc = React.useCallback(() => setCounter(R.inc), [setCounter])
 
+  const s3 = AWS.S3.use()
+
   // eslint-disable-next-line react-hooks/exhaustive-deps
   const validate = React.useCallback(
     cacheDebounce(async (name) => {
       if (name) {
-        const res = await req({
-          endpoint: '/package_name_valid',
-          method: 'POST',
-          body: { name },
+        const list = await requests.bucketListing({
```

Comment:
i think we should check for package existence here, not for data existence (either by querying the package index or listing the `.quilt/named_packages/$handle` dir)

Reply:
I don't know anything about this, please, provide more details

Reply:
see the contents of `.quilt/named_packages` dir: it has all the packages registered in the bucket and has the `$handle/$pointer_file` structure, so to check if the package exists in the bucket you should list objects under `.quilt/named_packages/$handle/` prefix -- if there are any, the package exists, otherwise it does not.

Reply:
I've added method requests.ensurePackageIsPresent, where i check for `.quilt/named_packages/package_name/latest`. `headObject(".quilt/named_packages/package_name")` (just directory, without package revision) suddenly doesn't work

Reply:
> I've added method requests.ensurePackageIsPresent, where i check for .quilt/named_packages/package_name/latest.

i think listing revisions is more reliable, bc the `latest` tag may be absent for some reason

> headObject(".quilt/named_packages/package_name") (just directory, without package revision) suddenly doesn't work

that's expected, bc head request only works for objects, and there's no such object, there's only prefix, so you should use `listObjectsV2` for that prefix (you can limit number of keys to 1 to just see if there's _anything_ under that prefix, no matter what)

Reply:
I changed `requests.ensurePackageIsPresent`. Now it uses `request.bucketListing` (which uses `listObjectsV2`). I just check if any file under `.named_packages/package_name`